### ISSUE TITLE
[Question] Is there a way to call setters automatically?

### ISSUE THREAD

**Issue by pdugas** [schmittjoh/serializer#890]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | no
| BC Break report? | no
| RFC?             | no

I'm looking for a way to distinguish between data posted to my Symfony 4 REST API that includes a property set to null as opposed to the property being omitted.  I thought I'd get tricky and use a baseclass with __get() and __set() methods that expose protected properties while updating an array with the names of properties that have been set. Turns out not to work as initially implemented.

I ended up with the code below that appears to be working. I'm not pleased having to add @Accessor annotations but I can live with it.  I discovered that if I use Symfony's stock serializer, they're not needed (obviously) and neither is the `DTO::__call()` method.  I'd rather stick with JMS Serializer but thought I'd ask if there's a better way to accomplish what I'm doing.

This is the baseclass for incoming _data transfer objects_ which I deserialize posted JSON into using FOSrest's body parameter converter.
```
class DTO
{
    private $flags = [];

    public function __set(string $name, $value)
    {
        if (property_exists($this, $name)) {
            $this->flags[$name] = true;
            $this->$name = $value;
        }
    }

    public function __get(string $name)
    {
        return property_exists($this, $name)
            ? $this->$name : null;
    }

    public function __isset(string $name)
    {
        return property_exists($this, $name)
            && array_key_exists($name, $this->flags);
    }

    public function __unset(string $name)
    {
        if (property_exists($this, $name)) {
            unset($this->flags[$name]);
            $this->$name = null;
        }
    }

    public function __call(string $name, array $args)
    {
        if ('set' === substr($name, 0, 3)) {
            if (empty($args)) {
                throw new \InvalidArgumentException('Missing value parameter');
            }
            $this->__set(lcfirst(substr($name, 3)), $args[0]);
            return $this;
        }
        if ('get' === substr($name, 0, 3)) {
            return $this->__get(lcfirst(substr($name, 3)));
        }
        throw new \BadMethodCallException('Method '.get_class($this).'::'.$name.' not found');
    }
}
```
Here's an example that derives from the base class.
```
class NamedDTO extends DTO
{
    /**
     * @JMS\Type("string")
     * @JMS\Accessor(getter="getName", setter="setName");
     * @JMS\Expose
     */
    protected $name;

    /**
     * @JMS\Type("string")
     * @JMS\Accessor(getter="getNote", setter="setNote");
     * @JMS\Expose
     */
    protected $note;
}
```

**Comment by goetas** [comment]
Personally I think this is just one approach to solve your very specific use case. I do not think that there is something wrong with it. It is just very specific for your need.

To be honest I would argue that the need to differentiate "not set" properties versus "null properties" vs "not present" properties is a bit weird. I would try to get rid of that constraint..... but this is more about the architecture of your application and I can argue with that.

Closing the issue for now, as it is not serializer related. The discussion can continue on a closed issue.

**Comment by holtkamp** [comment]
@pdugas did not read with a lot of attention, but for the hydration part you can also consider https://framework.zend.com/manual/2.4/en/modules/zend.stdlib.hydrator.html

**Comment by pdugas** [comment]
The desire to detect "null" vs "not set" goes along with handling PATCH operations to update existing objects. We need to support clearing a "lockedAt" property on a User entity for example by setting it to null. I won't argue there aren't other ways to do this but it's the approach we've taken.

The reason I chose to submit this was primarily the discovery that Symfony's native Serializer behaves as I expected while JMS Serializer doesn't. The undeniably clever approach JMS uses bypasses my __set() methods unless I add the @Accessor and __call bits. I guess I'm looking primarily for confirmation that this is by design and not a bug.

@holtkamp I'm certainly not an expert here but I thought traits were supposed to be used to solve multiple-inheritance issues. We chose to used a base class since there are no others.
